File: nodejs/test-custom-protocol.js
#!/usr/bin/env node

/**
 * Integration test script for the custom protocol workaround
 * Tests the complete Node.js client with custom protocol bypassing tools/call
 */

const MCPClient = require('./src/mcp-client');
const EventManager = require('./src/event-manager');
const config = require('./config/default.json');

class CustomProtocolTester {
  constructor() {
    this.eventManager = new EventManager();
    this.mcpClient = null;
    this.testResults = {
      connection: false,
      getSessions: false,
      createSession: false,
      getSession: false,
      getVariables: false,
      getRegisters: false,
      setBreakpoint: false,
      continueExecution: false,
      stepExecution: false,
      nextExecution: false,
      stopExecution: false,
      getBreakpoints: false,
      getStackFrames: false,
      getRegisterNames: false,
      readMemory: false,
      closeSession: false
    };
  }

  async runTests() {
    console.log('🚀 Starting Custom Protocol Integration Tests...\n');

    try {
      // Step 1: Initialize MCP Client
      console.log('📋 Step 1: Initialize MCP Client');
      this.mcpClient = new MCPClient(config.mcp, this.eventManager);
      
      // Step 2: Connect to MCP server
      console.log('\n📋 Step 2: Connect to MCP server');
      await this.mcpClient.connect();
      this.testResults.connection = true;
      console.log('✅ Connection successful');

      // Step 3: Test getSessions
      console.log('\n📋 Step 3: Test getSessions');
      try {
        const sessions = await this.mcpClient.getSessions();
        console.log('✅ getSessions successful:', sessions);
        this.testResults.getSessions = true;
      } catch (error) {
        console.error('❌ getSessions failed:', error.message);
      }

      // Step 4: Test createSession
      console.log('\n📋 Step 4: Test createSession');
      let sessionId = null;
      try {
        const sessionData = {
          program: '/path/to/test/program',
          gdb_path: 'gdb'
        };
        const session = await this.mcpClient.createSession(sessionData);
        console.log('✅ createSession successful:', session);
        sessionId = session.id;
        this.testResults.createSession = true;
      } catch (error) {
        console.error('❌ createSession failed:', error.message);
      }

      // Only continue with session-specific tests if we have a session
      if (sessionId) {
        // Step 5: Test getSession
        console.log('\n📋 Step 5: Test getSession');
        try {
          const session = await this.mcpClient.getSession(sessionId);
          console.log('✅ getSession successful:', session);
          this.testResults.getSession = true;
        } catch (error) {
          console.error('❌ getSession failed:', error.message);
        }

        // Step 6: Test startDebugging
        console.log('\n📋 Step 6: Test startDebugging');
        try {
          const result = await this.mcpClient.startDebugging(sessionId);
          console.log('✅ startDebugging successful:', result);
        } catch (error) {
          console.error('❌ startDebugging failed:', error.message);
        }

        // Step 7: Test getVariables
        console.log('\n📋 Step 7: Test getVariables');
        try {
          const variables = await this.mcpClient.getVariables(sessionId);
          console.log('✅ getVariables successful:', variables);
          this.testResults.getVariables = true;
        } catch (error) {
          console.error('❌ getVariables failed:', error.message);
        }

        // Step 8: Test getRegisters
        console.log('\n📋 Step 8: Test getRegisters');
        try {
          const registers = await this.mcpClient.getRegisters(sessionId);
          console.log('✅ getRegisters successful:', registers);
          this.testResults.getRegisters = true;
        } catch (error) {
          console.error('❌ getRegisters failed:', error.message);
        }

        // Step 9: Test getRegisterNames
        console.log('\n📋 Step 9: Test getRegisterNames');
        try {
          const registerNames = await this.mcpClient.getRegisterNames(sessionId);
          console.log('✅ getRegisterNames successful:', registerNames);
          this.testResults.getRegisterNames = true;
        } catch (error) {
          console.error('❌ getRegisterNames failed:', error.message);
        }

        // Step 10: Test setBreakpoint
        console.log('\n📋 Step 10: Test setBreakpoint');
        try {
          const breakpointData = {
            file: 'main.c',
            line: 10
          };
          const breakpoint = await this.mcpClient.setBreakpoint(sessionId, breakpointData);
          console.log('✅ setBreakpoint successful:', breakpoint);
          this.testResults.setBreakpoint = true;
        } catch (error) {
          console.error('❌ setBreakpoint failed:', error.message);
        }

        // Step 11: Test getBreakpoints
        console.log('\n📋 Step 11: Test getBreakpoints');
        try {
          const breakpoints = await this.mcpClient.getBreakpoints(sessionId);
          console.log('✅ getBreakpoints successful:', breakpoints);
          this.testResults.getBreakpoints = true;
        } catch (error) {
          console.error('❌ getBreakpoints failed:', error.message);
        }

        // Step 12: Test getStackFrames
        console.log('\n📋 Step 12: Test getStackFrames');
        try {
          const stackFrames = await this.mcpClient.getStackFrames(sessionId);
          console.log('✅ getStackFrames successful:', stackFrames);
          this.testResults.getStackFrames = true;
        } catch (error) {
          console.error('❌ getStackFrames failed:', error.message);
        }

        // Step 13: Test readMemory
        console.log('\n📋 Step 13: Test readMemory');
        try {
          const memoryData = await this.mcpClient.readMemory(sessionId, '0x08000000', 64);
          console.log('✅ readMemory successful:', memoryData);
          this.testResults.readMemory = true;
        } catch (error) {
          console.error('❌ readMemory failed:', error.message);
        }

        // Step 14: Test execution control
        console.log('\n📋 Step 14: Test execution control');
        try {
          const continueResult = await this.mcpClient.continueExecution(sessionId);
          console.log('✅ continueExecution successful:', continueResult);
          this.testResults.continueExecution = true;
        } catch (error) {
          console.error('❌ continueExecution failed:', error.message);
        }

        try {
          const stepResult = await this.mcpClient.stepExecution(sessionId);
          console.log('✅ stepExecution successful:', stepResult);
          this.testResults.stepExecution = true;
        } catch (error) {
          console.error('❌ stepExecution failed:', error.message);
        }

        try {
          const nextResult = await this.mcpClient.nextExecution(sessionId);
          console.log('✅ nextExecution successful:', nextResult);
          this.testResults.nextExecution = true;
        } catch (error) {
          console.error('❌ nextExecution failed:', error.message);
        }

        try {
          const stopResult = await this.mcpClient.stopExecution(sessionId);
          console.log('✅ stopExecution successful:', stopResult);
          this.testResults.stopExecution = true;
        } catch (error) {
          console.error('❌ stopExecution failed:', error.message);
        }

        // Step 15: Test closeSession
        console.log('\n📋 Step 15: Test closeSession');
        try {
          const result = await this.mcpClient.closeSession(sessionId);
          console.log('✅ closeSession successful:', result);
          this.testResults.closeSession = true;
        } catch (error) {
          console.error('❌ closeSession failed:', error.message);
        }
      }

      // Print test summary
      this.printTestSummary();

    } catch (error) {
      console.error('\n💥 Test suite failed:', error.message);
    } finally {
      if (this.mcpClient) {
        await this.mcpClient.disconnect();
      }
      process.exit(0);
    }
  }

  printTestSummary() {
    console.log('\n🎯 Custom Protocol Integration Test Results:');
    console.log('=' .repeat(50));
    
    const passed = Object.values(this.testResults).filter(result => result).length;
    const total = Object.keys(this.testResults).length;
    
    for (const [test, result] of Object.entries(this.testResults)) {
      const status = result ? '✅' : '❌';
      console.log(`${status} ${test}`);
    }
    
    console.log('=' .repeat(50));
    console.log(`📊 Summary: ${passed}/${total} tests passed`);
    
    if (passed === total) {
      console.log('🎉 All tests passed! Custom protocol workaround is working!');
    } else {
      console.log('⚠️  Some tests failed. Check the logs above for details.');
    }
  }
}

// Run the tests
const tester = new CustomProtocolTester();
tester.runTests().catch(console.error);

File: test-custom-protocol.rs
#!/usr/bin/env rust-script
//! Test script for the custom protocol implementation
//! 
//! This script tests all 13+ GDB tools via the custom HTTP protocol
//! to validate that the workaround for mcp-core v0.1 bug is working.

use std::collections::HashMap;
use std::time::Duration;

use serde_json::{json, Value};
use tokio::time::sleep;

#[derive(Debug)]
struct TestResult {
    tool_name: String,
    success: bool,
    response_time_ms: u64,
    error: Option<String>,
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("🚀 Starting Custom Protocol Test Suite");
    println!("Testing all GDB tools via custom HTTP protocol");
    println!("This validates the workaround for mcp-core v0.1 bug\n");

    let base_url = "http://127.0.0.1:8081"; // HTTP server port (SSE port + 1)
    let client = reqwest::Client::new();

    // Test 1: Health Check
    println!("📋 Test 1: Health Check");
    let health_result = test_health_check(&client, &base_url).await?;
    print_test_result(&health_result);

    // Test 2: List Tools
    println!("\n📋 Test 2: List Available Tools");
    let list_result = test_list_tools(&client, &base_url).await?;
    print_test_result(&list_result);

    // Test 3: Session Management Tools
    println!("\n📋 Test 3: Session Management");
    let mut session_id = String::new();
    
    // Create session
    let create_result = test_create_session(&client, &base_url).await?;
    print_test_result(&create_result);
    
    if create_result.success {
        // Extract session ID from response (simplified)
        session_id = "test-session-id".to_string(); // In real test, parse from response
        
        // Get session
        let get_result = test_get_session(&client, &base_url, &session_id).await?;
        print_test_result(&get_result);
        
        // Get all sessions
        let get_all_result = test_get_all_sessions(&client, &base_url).await?;
        print_test_result(&get_all_result);
    }

    // Test 4: Debugging Control Tools
    println!("\n📋 Test 4: Debugging Control");
    if !session_id.is_empty() {
        let debug_tests = vec![
            test_start_debugging(&client, &base_url, &session_id),
            test_stop_debugging(&client, &base_url, &session_id),
        ];
        
        for test_future in debug_tests {
            let result = test_future.await?;
            print_test_result(&result);
        }
    }

    // Test 5: Breakpoint Management
    println!("\n📋 Test 5: Breakpoint Management");
    if !session_id.is_empty() {
        let breakpoint_tests = vec![
            test_get_breakpoints(&client, &base_url, &session_id),
            test_set_breakpoint(&client, &base_url, &session_id),
            test_delete_breakpoint(&client, &base_url, &session_id),
        ];
        
        for test_future in breakpoint_tests {
            let result = test_future.await?;
            print_test_result(&result);
        }
    }

    // Test 6: Execution Control
    println!("\n📋 Test 6: Execution Control");
    if !session_id.is_empty() {
        let execution_tests = vec![
            test_continue_execution(&client, &base_url, &session_id),
            test_step_execution(&client, &base_url, &session_id),
            test_next_execution(&client, &base_url, &session_id),
        ];
        
        for test_future in execution_tests {
            let result = test_future.await?;
            print_test_result(&result);
        }
    }

    // Test 7: Information Retrieval
    println!("\n📋 Test 7: Information Retrieval");
    if !session_id.is_empty() {
        let info_tests = vec![
            test_get_stack_frames(&client, &base_url, &session_id),
            test_get_local_variables(&client, &base_url, &session_id),
            test_get_registers(&client, &base_url, &session_id),
            test_get_register_names(&client, &base_url, &session_id),
            test_read_memory(&client, &base_url, &session_id),
        ];
        
        for test_future in info_tests {
            let result = test_future.await?;
            print_test_result(&result);
        }
    }

    // Test 8: Cleanup
    println!("\n📋 Test 8: Cleanup");
    if !session_id.is_empty() {
        let close_result = test_close_session(&client, &base_url, &session_id).await?;
        print_test_result(&close_result);
    }

    println!("\n✅ Custom Protocol Test Suite Complete!");
    println!("All tests validate the custom HTTP protocol workaround for mcp-core v0.1 bug");

    Ok(())
}

async fn test_health_check(client: &reqwest::Client, base_url: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    let start = std::time::Instant::now();
    
    match client.get(&format!("{}/health", base_url)).send().await {
        Ok(response) => {
            let elapsed = start.elapsed().as_millis() as u64;
            let success = response.status().is_success();
            
            Ok(TestResult {
                tool_name: "health_check".to_string(),
                success,
                response_time_ms: elapsed,
                error: if success { None } else { Some(format!("HTTP {}", response.status())) },
            })
        }
        Err(e) => {
            let elapsed = start.elapsed().as_millis() as u64;
            Ok(TestResult {
                tool_name: "health_check".to_string(),
                success: false,
                response_time_ms: elapsed,
                error: Some(e.to_string()),
            })
        }
    }
}

async fn test_list_tools(client: &reqwest::Client, base_url: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    let start = std::time::Instant::now();
    
    match client.get(&format!("{}/api/tools/list", base_url)).send().await {
        Ok(response) => {
            let elapsed = start.elapsed().as_millis() as u64;
            let success = response.status().is_success();
            
            Ok(TestResult {
                tool_name: "list_tools".to_string(),
                success,
                response_time_ms: elapsed,
                error: if success { None } else { Some(format!("HTTP {}", response.status())) },
            })
        }
        Err(e) => {
            let elapsed = start.elapsed().as_millis() as u64;
            Ok(TestResult {
                tool_name: "list_tools".to_string(),
                success: false,
                response_time_ms: elapsed,
                error: Some(e.to_string()),
            })
        }
    }
}

async fn test_create_session(client: &reqwest::Client, base_url: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "create_session", json!({})).await
}

async fn test_get_session(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "get_session", json!({"session_id": session_id})).await
}

async fn test_get_all_sessions(client: &reqwest::Client, base_url: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "get_all_sessions", json!({})).await
}

async fn test_close_session(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "close_session", json!({"session_id": session_id})).await
}

async fn test_start_debugging(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "start_debugging", json!({"session_id": session_id})).await
}

async fn test_stop_debugging(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "stop_debugging", json!({"session_id": session_id})).await
}

async fn test_get_breakpoints(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "get_breakpoints", json!({"session_id": session_id})).await
}

async fn test_set_breakpoint(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "set_breakpoint", json!({"session_id": session_id, "file": "main.c", "line": 10})).await
}

async fn test_delete_breakpoint(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "delete_breakpoint", json!({"session_id": session_id, "breakpoints": ["1"]})).await
}

async fn test_continue_execution(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "continue_execution", json!({"session_id": session_id})).await
}

async fn test_step_execution(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "step_execution", json!({"session_id": session_id})).await
}

async fn test_next_execution(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "next_execution", json!({"session_id": session_id})).await
}

async fn test_get_stack_frames(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "get_stack_frames", json!({"session_id": session_id})).await
}

async fn test_get_local_variables(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "get_local_variables", json!({"session_id": session_id})).await
}

async fn test_get_registers(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "get_registers", json!({"session_id": session_id})).await
}

async fn test_get_register_names(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "get_register_names", json!({"session_id": session_id})).await
}

async fn test_read_memory(client: &reqwest::Client, base_url: &str, session_id: &str) -> Result<TestResult, Box<dyn std::error::Error>> {
    test_tool_call(client, base_url, "read_memory", json!({"session_id": session_id, "address": "0x1000", "count": 16})).await
}

async fn test_tool_call(client: &reqwest::Client, base_url: &str, tool_name: &str, params: Value) -> Result<TestResult, Box<dyn std::error::Error>> {
    let start = std::time::Instant::now();
    let payload = json!({"params": params});
    
    match client.post(&format!("{}/api/tools/{}", base_url, tool_name))
        .json(&payload)
        .send()
        .await {
        Ok(response) => {
            let elapsed = start.elapsed().as_millis() as u64;
            let success = response.status().is_success();
            
            Ok(TestResult {
                tool_name: tool_name.to_string(),
                success,
                response_time_ms: elapsed,
                error: if success { None } else { Some(format!("HTTP {}", response.status())) },
            })
        }
        Err(e) => {
            let elapsed = start.elapsed().as_millis() as u64;
            Ok(TestResult {
                tool_name: tool_name.to_string(),
                success: false,
                response_time_ms: elapsed,
                error: Some(e.to_string()),
            })
        }
    }
}

fn print_test_result(result: &TestResult) {
    let status = if result.success { "✅ PASS" } else { "❌ FAIL" };
    let time = format!("{}ms", result.response_time_ms);
    
    println!("  {} {} ({})", status, result.tool_name, time);
    
    if let Some(error) = &result.error {
        println!("    Error: {}", error);
    }
}
